SignalMessaging/attachments/ApproveContactShareViewController.swift

```diff
@@ -0,0 +1,556 @@
+        previewView.autoPinBottomToSuperviewMargin()
+        previewView.autoPinLeading(toTrailingEdgeOf: checkbox, offset: hSpacing)
+        previewView.autoPinTrailingToSuperviewMargin(withInset: hMargin)
+    }
+
+    func wasTapped(sender: UIGestureRecognizer) {
+        Logger.info("\(self.logTag) \(#function)")
+
+        guard sender.state == .recognized else {
+            return
+        }
+        field.setIsIncluded(!field.isIncluded())
+        checkbox.isSelected = field.isIncluded()
+    }
+}
+
+// MARK: -
+
+@objc
+public class ApproveContactShareViewController: OWSViewController, EditContactShareNameViewControllerDelegate {
+    weak var delegate: ApproveContactShareViewControllerDelegate?
+
+    let contactsManager: OWSContactsManager
+
+    var contactShare: OWSContact
+
+    var fieldViews = [ContactShareFieldView]()
+
+    var nameLabel: UILabel!
+
+    // MARK: Initializers
+
+    @available(*, unavailable, message:"use other constructor instead.")
+    required public init?(coder aDecoder: NSCoder) {
+        fatalError("unimplemented")
+    }
+
+    @objc
+    required public init(contactShare: OWSContact, contactsManager: OWSContactsManager, delegate: ApproveContactShareViewControllerDelegate) {
+        self.contactsManager = contactsManager
+        self.contactShare = contactShare
+        self.delegate = delegate
+
+        super.init(nibName: nil, bundle: nil)
+
+        buildFields()
+    }
+
+    func buildFields() {
+        var fieldViews = [ContactShareFieldView]()
+
+        // TODO: Avatar
+
+        for phoneNumber in contactShare.phoneNumbers {
+            let field = ContactSharePhoneNumber(phoneNumber)
+            let fieldView = ContactShareFieldView(field: field, previewViewBlock: { [weak self] _ in
+                guard let strongSelf = self else { return UIView() }
+                return strongSelf.previewView(forPhoneNumber: phoneNumber)
+            })
+            fieldViews.append(fieldView)
+        }
+        for email in contactShare.emails {
+            let field = ContactShareEmail(email)
+            let fieldView = ContactShareFieldView(field: field, previewViewBlock: { [weak self] _ in
+                guard let strongSelf = self else { return UIView() }
+                return strongSelf.previewView(forEmail: email)
+            })
+            fieldViews.append(fieldView)
+        }
+        for address in contactShare.addresses {
+            let field = ContactShareAddress(address)
+            let fieldView = ContactShareFieldView(field: field, previewViewBlock: { [weak self] _ in
+                guard let strongSelf = self else { return UIView() }
+                return strongSelf.previewView(forAddress: address)
+            })
+            fieldViews.append(fieldView)
+        }
+
+        self.fieldViews = fieldViews
+    }
+
+    override public var canBecomeFirstResponder: Bool {
+        return true
```

Comment:
Why do we want to become first responder?

Reply:
We still haven't fixed all of the issues around the input toolbar showing up, so it wasn't clear to me if we are going to need this solution too.  Removed for now.